src/types/bytearray.rs

```diff
@@ -21,6 +21,39 @@ impl PyByteArray {
         unsafe { py.from_owned_ptr::<PyByteArray>(ffi::PyByteArray_FromStringAndSize(ptr, len)) }
     }
 
+    /// Creates a new Python bytearray object.
+    /// The bytearray is zero-initialised and can be read inside `init`.
+    /// The `init` closure can further initialise the bytearray.
+    ///
+    /// Panics if out of memory.
+    ///
+    /// # Example
+    /// ```
+    /// use pyo3::{prelude::*, types::PyByteArray};
+    /// Python::with_gil(|py| {
+    ///     let py_bytearray = PyByteArray::new_with(py, 10, |bytes: &mut [u8]| {
+    ///         bytes.copy_from_slice(b"Hello Rust");
+    ///     });
+    ///     let bytearray: &[u8] = unsafe { py_bytearray.as_bytes() };
+    ///     assert_eq!(bytearray, b"Hello Rust");
+    /// });
+    /// ```
+    pub fn new_with<F: Fn(&mut [u8])>(py: Python<'_>, len: usize, init: F) -> &PyByteArray {
```

Comment:
I think we can make a stronger guarantee here (that the closure will only be called once).

```suggestion
    pub fn new_with<F>(py: Python, len: usize, init: F) -> &PyByteArray
    where
        F: FnOnce(&mut [u8])
    {
```